module: add support for abi stable module API

jasongin opened the pull request:
Add support for abi stable module API (N-API) as "Experimental feature".
The goal of this API is to provide a stable Node API for native
module developers. N-API aims to provide ABI compatibility guarantees
across different Node versions and also across different
Node VMs - allowing N-API enabled native modules to just work
across different versions and flavors of Node.js without recompilation.

A more detailed introduction is provided in:
https://github.com/nodejs/node-eps/blob/master/005-ABI-Stable-Module-API.md
and https://github.com/nodejs/abi-stable-node/blob/doc/VM%20Summit.pdf.

The feature, during its experimental state, will be guarded by a runtime
flag "--napi-modules". Only when this flag is added to the command line
will N-API modules along with regular non N-API modules be supported.

The API is defined by the methods in "src/node_api.h",
"src/node_api_types.h" and "src/node_api_async.h". This is the best
starting point to review the API surface. More documentation will follow.

In addition to the implementation of the API using V8, which is included
in this PR, the API has also been validated against chakracore and that
port is available in
https://github.com/nodejs/abi-stable-node/tree/api-prototype-chakracore-8.x.

The current plan is to provide N-API support in versions 8.X and 6.X
directly. For older versions, such as 4.X or pre N-API versions of 6.X,
we plan to create an external npm module to provide a migration path
that will allow modules targeting older Node.js versions to use the API,
albeit without getting the advantage of not having to recompile.

In addition, the external npm package will include C++ sugar to
simplify the use of the API. The sugar will be in-line only and will
only use the exported N-API methods but is not part of the N-API
itself. The current version is in:
https://github.com/nodejs/node-api.

This PR is a result of work in the abi-stable-node repo:
https://github.com/nodejs/abi-stable-node/tree/doc,
with this PR being the cumulative work on the api-prototype-8.x
branch with the following contributors in alphabetical order:

 Arunesh Chandra (@aruneshchandra)
 Gabriel Schulhof (@gabrielschulhof)
 Hitesh Kanwathirtha (@digitalInfinity)
 Ian Halliday (@ianwjhalliday)
 Jason Ginchereau (@jasongin)
 Michael Dawson (@mhdawson)
 Sampson Gao (@sampsongao)
 Taylor Woll (@boingoing)

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added - IN PROGRESS
- [x] commit message follows [commit guidelines][]

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
module

[commit guidelines]: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines

richardlau commented on src/node.cc:
```diff
@@ -3535,6 +3547,7 @@ static void PrintHelp() {
          "  --trace-deprecation        show stack traces on deprecations\n"
          "  --throw-deprecation        throw an exception on deprecations\n"
          "  --no-warnings              silence all process warnings\n"
+         "  --napi-modules[=yes|no]    load N-API modules (default no)\n"
```
As the comment at the beginning of this function says:
```C++
   // XXX: If you add an option here, please also add it to doc/node.1 and
   // doc/api/cli.md
```

jasongin replied:
> If somebody really wants to avoid loading N-API modules even after we've eventually set the default to "yes"

I'm not sure that's something that we'll need to support at that point. If N-API is successful it should become the *normal* way of writing native modules, so it wouldn't really make sense to disable it.

mhdawson replied:
I agree. Once non-experimental we would would most likely just remove the flag.  I think at this point we could just likely make it --napi-modules with no options.

Fishrock123 replied:
I personally would prefer `--napi-modules` and then once it goes live we could add an optional `=false` or something to force-disable?

jasongin replied:
@Fishrock123 can you explain why you (or someone) might want to disable the feature after it is out of experimental status?

Fishrock123 replied:
No, actually. I really do not know, but I would prefer to not have `=yes`.

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/87c42e70fcdb4247c79f1a07c6cc335c9ff64287

trevnorris commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  RETURN_STATUS_IF_FALSE(v8impl::TryCatch::lastException().IsEmpty(), \
+                         napi_pending_exception);                     \
+  napi_clear_last_error();                                            \
+  v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)))
```
Looks like `NAP_PREAMBLE` is used in many functions that don't execute javascript. what's the rational for using TryCatch when no JS is called and the v8 Maybe APIs are used?

boingoing replied:
We have made an effort to avoid using ```NAPI_PREAMBLE``` for functions which don't execute JavaScript (see ```napi_is_error``` for example) but we haven't done that exhaustively. I think we took the more conservative approach of including TryCatch when we weren't sure if it was strictly required or not because the guidance seems to be that using TryCatch was cheap enough that this would be fine. If we're sure no JavaScript is getting executed, removing the ```NAPI_PREAMBLE``` seems like the right thing to do.

jasongin replied:
Also we know we need to remove the `NAPI_PREAMBLE` from at least a few of the functions, for https://github.com/nodejs/abi-stable-node/issues/122 

trevnorris replied:
Thanks for the explanation. To clarify, I don't see this as a blocker.

trevnorris commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  } else if (finalize_cb != nullptr) {
+    // Create a self-deleting reference just for the finalize callback.
+    new v8impl::Reference(
+        isolate, obj, 0, true, finalize_cb, nativeObj, finalize_hint);
```
Not sure I can appreciate this auto-magical deletion of the the pointer. One reason is because there's no way to know whether the pointer was made via `new` or `malloc`. Even in `node_base_object.h` there's an explicit call to make the object weak.

trevnorris replied:
Edit: missed the "self-deleting" part. Nothing to see here.

trevnorris commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+    }
+
+    if (deleteSelf) {
+      delete reference;
```
It feels counter intuitive to go to such lengths to make the public API be C, but then use `delete` on a user-supplied pointer.

boingoing replied:
The ```delete``` here is deleting our ```Reference``` instance - not the user-supplied pointer. We aren't (hopefully) doing anything which would preclude a module written in C.

boingoing replied:
The ```delete``` here is deleting our ```Reference``` instance - not the user-supplied pointer. We aren't (hopefully) doing anything which would preclude a module written in C.

There is another ```delete reference``` in ```napi_delete_reference``` which operates on a user-supplied pointer (the ```napi_ref``` argument). We are expecting the user to hand us back a ```Reference*``` which we allocated (via ```new```) in ```napi_create_reference```. This ```Reference``` class acts as a container for us to do ref-counting.

If the user code calls ```napi_delete_reference``` with any random pointer, I guess behavior would be undefined. We would try to use ```~Reference``` which might crash. But this is true of all the API which take an opaque pointer and cast it to an engine-specific type. We have a similar pattern for ```napi_close_handle_scope``` and ```napi_close_escapable_handle_scope```.

gabrielschulhof replied:
@trevnorris it sounds like we should have a test where an addon is implemented in C, just to keep us honest.

jasongin replied:
https://github.com/nodejs/node/pull/11975/commits/fb8ced493fe5313a2fd082bd6f23d24b268b644f adds `New()` and `Delete()` methods on the `Reference` class for clarity.

trevnorris replied:
my bad. misunderstood what was going on here.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+}
+
+napi_status napi_escape_handle(napi_env e,
+       
```
Did you mean to put this comment somewhere else?

I'm guessing you might be referring to the access of the `prototype` property in `napi_instanceof()` [here](https://github.com/nodejs/node/pull/11975/files/43ab5a8b03d05c93054bbdb264a56bae0881523b#diff-643397d9aacb4067c9937f6d6dfdbef2R2059). In that case, it is really a function's `prototype` property that is needed, that is different from the `__proto__` property that is returned by `GetPrototype()`. The "instanceof" logic had to be implemented in this way rather than using `v8::FunctionTemplate::HasInstance()` because N-API does not expose the V8-specific concept of function or object templates, nor does it track them internally, so the `FunctionTemplate` that was used to create the function passed to `napi_instanceof()` is not available at that point.

(There is also a `napi_get_prototype()` API that simply calls `v8::Object::GetPrototype()`.)

addaleax replied:
It seems to me like `int` is a V8 implementation detail, `size_t` might better for argument counts (here and elsewhere)?

addaleax replied:
`v8::Handle` → `v8::Local` (also something for search & replace… :smile:)

addaleax replied:
If we want to support an `instanceof` implementation that doesn’t conform to ES6 (in that it isn’t based on `[Symbol.hasInstance]()`), could/should we call it something else? Alternatively, could this wrapper just manually call out to `constructor[Symbol.hasInstance](object)`?

jasongin replied:
This does conform to ES6 `instanceof`. There are extensive tests at `test/addons-napi/test_instanceof` that re-run all of V8's `instanceof` tests using this API.

gabrielschulhof replied:
@addaleax I implemented `napi_instanceof()` this way because I didn't have access to the implementation of the operator. However, if v8 provides access to an implementation that does not require retaining/acquiring the `FunctionTemplate` then I'd be happy to change this API. Additionally, [it looks like](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/hasInstance) `Symbol.hasInstance()` allows users to customize the behaviour of the `instanceof` operator, so we absolutely have to use `Symbol.hasInstance()` when available.

addaleax replied:
@jasongin Yeah, it matches [`OrdinaryHasInstance()`](https://tc39.github.io/ecma262/#sec-ordinaryhasinstance). I’m good with just naming it that.

Otoh, it should really not be a problem for V8 to expose `InstanceOf()` and `OrdinaryHasInstance()` directly…

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/9e3ab8317b107ecd4456aecf302d7ed4fb8af2e9

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/116bd1945e0204e10600ddc6c87cf9632eff477c

trevnorris replied:
ah yup. see what you're saying.

addaleax commented:
Btw, in ba4847e879424ad173289e8fb96cc86a09ee899b we listed the authors as `Author: name <email>` as PR metadata, it looks like doing the same thing might make sense here, too?

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(v8::Exception::Error(
+      v8impl::V8LocalValueFromJsValue(msg).As<v8::String>()));
```
Can/should these include actual checks to make sure that `msg` does refer to a string?

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/6b4ce5389dbb9642ae5f22ecc2cb2c066913d451

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  } else if (v->IsExternal()) {
+    *result = napi_external;
+  } else {
+    *result = napi_object;  // Is this correct?
```
I don’t quite understand this line/comment… if `v->IsObject()` is false, why return `napi_object`? Also, I think the above list is exhaustive.

jasongin replied:
It should return an error in this case - Resolved by https://github.com/nodejs/node/pull/11975/commits/6b4ce5389dbb9642ae5f22ecc2cb2c066913d451

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  CHECK_ARG(result);
+
+  v8::Local<v8::Value> val = v8impl::V8LocalValueFromJsValue(value);
+  RETURN_STATUS_IF_FALSE(val->IsNumber(), napi_number_expected);
```
`IsUint32`?

jasongin replied:
`IsNumber()` is actually desired here, because the following `Value.As<Int32>()` works when the Value is any kind of Number. (And same for casting to the other Number types with `As`.)

https://github.com/nodejs/node/pull/11975/commits/1b2f2db541d555cf5f4087c5ed79c7a4f8c419ac adds comments here about that along with test cases to validate all the permutations of type casts and type coercions.

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  v8::Local<v8::Value> val = v8impl::V8LocalValueFromJsValue(value);
+  RETURN_STATUS_IF_FALSE(val->IsNumber(), napi_number_expected);
+
+  *result = val.As<v8::Integer>()->Value();
```
Can you just cast this to `Integer`, even if you only checked that `IsNumber()` is true?

jasongin replied:
Yes, see tests added in https://github.com/nodejs/node/pull/11975/commits/1b2f2db541d555cf5f4087c5ed79c7a4f8c419ac

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+// Gets the number of CHARACTERS in the string.
+napi_status napi_get_value_string_length(napi_env e,
+                                         napi_value value,
+                                         int* result) {
```
It seems like `int` is a V8 implementation detail here too, `size_t` is probably better for string lengths (also here and below)?

jasongin replied:
Yes, good point.

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/9e3ab8317b107ecd4456aecf302d7ed4fb8af2e9

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  CHECK_ARG(ref);
+
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  delete reference;
```
(re: my comment above … maybe also make the `Reference` destructor private and add a `Delete` static? I know it might seem a bit superfluous but I think it would really help with readability)

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/fb8ced493fe5313a2fd082bd6f23d24b268b644f

jasongin commented on src/node_api.h:
```diff
@@ -0,0 +1,483 @@
+NAPI_EXTERN napi_status napi_create_string_utf16(napi_env env,
+                                                 const char16_t* s,
+                                                 int length,
+                                                 napi_value* result);
```
Good feedback, we should consider adding it. I opened https://github.com/nodejs/abi-stable-node/issues/179 to track this.

Note *missing* APIs should probably not block this PR unless it is something critical. The goal for this PR is to establish an initial set of APIs that provide enough functionality for native module developers to begin experimenting and give us feedback. We have plans to expand on these APIs in the coming months, before the feature graduates out of experimental status.

mhdawson replied:
As @jasongin mentioned, we don't believe missing APIs should be a blocker.  There are a few others we have in our plans to add which we will prioritized based on feedback from people using the API.

jasongin commented on src/node.cc:
```diff
@@ -4473,6 +4492,11 @@ inline int Start(Isolate* isolate, IsolateData* isolate_data,
+    ProcessEmitWarning(&env, "N-API is an experimental feature "
+        "and could change at any time.");
+  }
+
```
CTC has specifically requested command-line enablement along with a warning message for features with experimental status. Tracing does the same thing.

mikeal replied:
Is there a link somewhere to that decision?

I don't quite understand the justification for a printed warning for a feature that can't even be used without a flag. If you're using the flag, you wanted the feature and are quite aware that it is experimental. Do any V8 features do this?

aruneshchandra replied:
VM Summit 3 [meeting notes](https://github.com/joshgav/abi-stable-node/blob/f7fc7e4bddd222aba922d52f4447bcab2fd40691/meeting%20notes/2017-03-03.md) has this 
_For 8.0, include an opt-in runtime flag, show experimental warning when a NAPI module is loaded._

jasongin replied:
If consensus is that _documentation_ of experimental status is sufficient and this warning message is not necessary, we can take it out. I don't recall anyone feeling very strongly about this at the VM summit.

addaleax replied:
@jasongin I’d say take it out, and we can re-visit that in the unlikely case anybody complains.

mhdawson replied:
I think the main motivation was to ensure that anybody turning it on is fully aware of the status and to be consistent with what we have done for other experimental features.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+    return _refcount;
+  }
+
+  int Release() {
```
Sure, that probably makes sense here... but note this code is just private implementation details.

More generally, we explicitly do not have a goal to project V8 concepts and terminology into the public API surface, except where it matches standard JavaScript concepts and terminology. This is meant to be a Node.js API, *not* a V8 API. A developer familiar with JavaScript should be able to understand the API without having any knowledge (or documentation) of V8.

addaleax replied:
@jasongin Exactly, the other uses in Node are also somewhat internal-y. And also, I agree, it’s obviously good to have an engine-agnostic API – I don’t think V8 actually exposes anything like this anyway, though. :)

jasongin replied:
Ahh you're right, sorry I got a little mixed up. The public [`node::ObjectWrap`](https://github.com/nodejs/node/blob/master/src/node_object_wrap.h) does indeed use `Ref/Unref`, so we should probably follow that.

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/fb8ced493fe5313a2fd082bd6f23d24b268b644f

kkoopa commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+    U(napi_value _v) : v(_v) {}
+  } u(v);
+  assert(sizeof(u.v) == sizeof(u.l));
+  return u.l;
```
There already is one, but it is private.

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/116bd1945e0204e10600ddc6c87cf9632eff477c

I couldn't use `std::is_trivially_copyable` because it wasn't available with some older compilers that are still supported by Node.js.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  } u(local);
+  assert(sizeof(u.v) == sizeof(u.l));
+  return u.v;
+}
```
> return reinterpret_cast<napi_value>(*local)

Yes, I agree that's correct and more straightforward. I'll update it.

(This is some of the oldest code in the N-API project; I'm guessing whoever wrote it was initially trying to avoid using `reinterpret_cast<>`, though the union accomplishes the same thing in a roundabout way.)

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/116bd1945e0204e10600ddc6c87cf9632eff477c

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  /*virtual*/
+  void SetReturnValue(napi_value v) override {
+    // Cannot set the return value of a setter.
+    assert(false);
```
Changing to `node::FatalError()`. (It calls `ABORT()` internally.)

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/116bd1945e0204e10600ddc6c87cf9632eff477c

boingoing commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  *result = v8impl::JsValueFromV8LocalValue(scope.Escape(tpl->GetFunction()));
+
+  if (staticPropertyCount > 0) {
+    std::vector<napi_property_descriptor> staticDescriptors;
```
That's fair. We haven't been very strict about consistency here. Couple questions before I dig in and rename all of these. :)

- I assume snake_case should also be used for all parameters, function names, etc? IIRC, some parameter names are camelCased right now.
- Are there rules around compound words? Is ```int bufsize``` allowed, for example, or should it always be ```int buf_size``` or ```int buf_length```?
- Leading underscores in names like ```_this```? 
- Are there casing rules for classes and class methods? For example, should we rename ```CallbackWrapper``` or ```CallbackWrapper::SetReturnValue``` or ```CallbackWrapper::_argsLength```?

addaleax replied:
I’m mostly echoing what @bnoordhuis would mention anyway, so pinging him to answer your questions ;)

Personally, I’m okay with `bufsize`, and I think classes and methods consistently use CamelCase.

jasongin replied:
Resolved by https://github.com/nodejs/node/pull/11975/commits/fb8ced493fe5313a2fd082bd6f23d24b268b644f which changes locals and parameter names to snake_case.

jasongin commented on src/node.cc:
```diff
@@ -2463,15 +2466,24 @@ void DLOpen(const FunctionCallbackInfo<Value>& args) {
+      snprintf(errmsg,
+               sizeof(errmsg),
+               "The module '%s'"
+               "\nwas compiled against the Node.js API. This feature is "
```
Resolved by https://github.com/nodejs/node/pull/11975/commits/b12962494452c2cc95d55dc79d017e0db517c746

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+    v8::Local<v8::Value> l;
+    U(v8::Local<v8::Value> _l) : l(_l) {}
+  } u(local);
+  assert(sizeof(u.v) == sizeof(u.l));
```
Resolved by https://github.com/nodejs/node/pull/11975/commits/116bd1945e0204e10600ddc6c87cf9632eff477c

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  void* _data;
+};
+
+template <typename T, int I>
```
Resolved by https://github.com/nodejs/node/pull/11975/commits/116bd1945e0204e10600ddc6c87cf9632eff477c

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+
+const napi_extended_error_info* napi_get_last_error_info() {
+  static_assert(
+      sizeof(error_messages) / sizeof(*error_messages) == napi_status_last,
```
Resolved by https://github.com/nodejs/node/pull/11975/commits/116bd1945e0204e10600ddc6c87cf9632eff477c

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+                              const char* utf8name,
+                              napi_callback constructor,
+                              void* data,
+                              int property_count,
```
Resolved by https://github.com/nodejs/node/pull/11975/commits/9e3ab8317b107ecd4456aecf302d7ed4fb8af2e9

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+  tpl->SetClassName(namestring);
+
+  int staticPropertyCount = 0;
+  for (int i = 0; i < property_count; i++) {
```
Resolved by https://github.com/nodejs/node/pull/11975/commits/9e3ab8317b107ecd4456aecf302d7ed4fb8af2e9 (`size_t`) and https://github.com/nodejs/node/pull/11975/commits/fb8ced493fe5313a2fd082bd6f23d24b268b644f (snake_case)

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+    std::vector<napi_property_descriptor> staticDescriptors;
+    staticDescriptors.reserve(staticPropertyCount);
+
+    for (int i = 0; i < property_count; i++) {
```
Resolved by https://github.com/nodejs/node/pull/11975/commits/9e3ab8317b107ecd4456aecf302d7ed4fb8af2e9

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2430 @@
+                              napi_value value) {
+  NAPI_PREAMBLE(e);
+
+  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
```
Fixed everywhere by https://github.com/nodejs/node/pull/11975/commits/116bd1945e0204e10600ddc6c87cf9632eff477c

jasongin commented:
Thanks all for the feedback. I just pushed a set of commits from @gabrielschulhof, @boingoing, and myself, which address *most* of the feedback so far.

We are still working on a couple things; these will be relatively isolated changes and should not block further review of the PR:
 - Improvement to `napi_instanceof`: https://github.com/nodejs/abi-stable-node/issues/182
 - Latin1 string support: https://github.com/nodejs/abi-stable-node/issues/179

Please keep the feedback coming!

jasongin commented:
@addaleax @trevnorris @richardlau We have addressed your feedback in updates to this PR. Would you mind taking another look? Thanks!

addaleax commented on src/node_api_types.h:
```diff
@@ -0,0 +1,94 @@
+typedef struct napi_callback_info__ *napi_callback_info;
+
+typedef void (*napi_callback)(napi_env, napi_callback_info);
+typedef void (*napi_finalize)(void* finalize_data, void* finalize_hint);
```
I find the names `finalize_data` and `finalize_hint` a bit nondescriptive, how about `target` for the former and `opaque` or something similar for the latter (assuming I’m understanding them correctly here)?

jasongin replied:
I think these names were inspired by "data" and "hint" parameters in `node::Buffer::FreeCallback`. (We were looking for some terms other than "context", which has other meanings.)

I'm not sure I agree that `target` and `opaque` are any more descriptive though.

addaleax replied:
@jasongin Ah, that makes sense. And agreed that “context” is not a good choice…

I don’t think I know better names either, I guess we’ll just need to have clear documentation.

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2465 @@
+    }
+  }
+
+  static v8::Persistent<v8::Value>& LastException() { return _the_exception; }
```
By the way, I’d be glad if we could avoid globals and instead turn these into per-Isolate/Environment properties… that shouldn’t be a blocker for this PR, though

jasongin replied:
Yes, good point. Not fixing for this PR, but I opened https://github.com/nodejs/abi-stable-node/issues/198 to track for later.

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2465 @@
+    cbdata->SetInternalField(
+        v8impl::kDataIndex,
+        v8::External::New(isolate, reinterpret_cast<void*>(data)));
+  }
```
This cannot be conditional, otherwise the corresponding `GetInternalField()` returns `undefined`, and in debug mode converting to an `v8::External` crashes. (Or, alternatively, you need to add a `IsExternal()` check in the `CallbackWrapperBase` constructor). Ditto for the accessor callbacks.

addaleax replied:
Also, just so you know, we prefer an explicit `if (foo != nullptr)` over just `if (foo)`

jasongin replied:
This seems to work fine. We have tests that cover this code, some of which use null callback data, and they are all passing in debug mode.

I even stepped through it in a debugger. After skipping the call to `SetInternalField()`, later on when the callback is invoked then this code:
```C++
v8::Local<v8::External>::Cast(_cbdata->GetInternalField(kDataIndex))->Value()
```
returns `nullptr`, and does not crash in debug mode.

jasongin replied:
Now I'm trying to find out why `V8_ENABLE_CHECKS` is not on in a debug build, and how to turn it on for testing.

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/f21ab80825e62b302da65abbea49cff834650668
Also updated to `if (foo != nullptr)` style in several places.

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2465 @@
+
+  // Value.As<Int32> works when the Value is any kind of Number.
+  RETURN_STATUS_IF_FALSE(val->IsNumber(), napi_number_expected);
+  *result = val.As<v8::Int32>()->Value();
```
(This is weird, the comments are shifting around? Github is being weird I guess…)

Anyway, this crashes in debug mode. You will need `IsUint32()`/`IsInt32()` checks here in order to use `.As<Int32>()` or `.As<Uint32>()`, `IsNumber()` is not enough (except for `.As<Integer>()`).

Alternatively, you could use `.As<Integer>()` and then truncate… not sure what the best behaviour is.

jasongin replied:
Are you sure? I just re-confirmed that the test covering this code (`test/addons-napi/test_conversions`) is passing in debug mode.

jasongin replied:
Actually what do you mean by "debug" mode? Is that a debug build, or when a debugger is attached?

jasongin replied:
OK I see these checks are enabled by the `V8_ENABLE_CHECKS` preprocessor define. But that doesn't get defined when I do a normal debug build.

addaleax replied:
How are you doing debug builds? I can definitely reproduce this by running

```sh
./configure --debug
make -j8
/usr/bin/python tools/test.py --mode=debug addons-napi
```

jasongin replied:
I've mostly been working on Windows, so I run:
```cmd
.\vcbuild.bat debug x64 test-addons-napi
```

The `node.vcxproj` generated from `node.gyp` has no reference to `V8_ENABLE_CHECKS`, although the `*.vcxproj` files under `deps/v8` do.

I have a build in progress on a Mac, so I'll try to reproduce the issue there.

jasongin replied:
Ahh, I do see the fatal errors from casting on a debug build on Mac. I'll fix those.

So apparently there is a problem with `V8_ENABLE_CHECKS` not getting enabled on Windows debug builds, _unrelated to this PR_. I'll try to look into that, but I'm not an expert on the build system.

addaleax replied:
@nodejs/v8 @nodejs/platform-windows ^^^

jasongin replied:
This is interesting... I checked the compiler flags used when building on Mac. I don't see `V8_ENABLE_CHECKS` being defined there either, when compiling the node core sources in debug build mode.

> c++ '-D_DARWIN_USE_64_BIT_INODE=1' '-DNODE_ARCH="x64"' '-DNODE_WANT_INTERNALS=1' '-DV8_DEPRECATION_WARNINGS=1' '-DNODE_USE_V8_PLATFORM=1' '-DNODE_HAVE_I18N_SUPPORT=1' '-DNODE_HAVE_SMALL_ICU=1' '-DHAVE_INSPECTOR=1' '-DHAVE_OPENSSL=1' '-DHAVE_DTRACE=1' '-D__POSIX__' '-DNODE_PLATFORM="darwin"' '-DUCONFIG_NO_SERVICE=1' '-DUCONFIG_NO_REGULAR_EXPRESSIONS=1' '-DU_ENABLE_DYLOAD=0' '-DU_STATIC_IMPLEMENTATION=1' '-DU_HAVE_STD_STRING=0' '-DUCONFIG_NO_BREAK_ITERATION=0' '-DUCONFIG_NO_LEGACY_CONVERSION=1' '-DHTTP_PARSER_STRICT=0' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DDEBUG' '-D_DEBUG' -I../src -I../tools/msvs/genfiles -I../deps/uv/src/ares -I/Users/jasongin/src/napi/napi-8/out/Debug/obj/gen -I/Users/jasongin/src/napi/napi-8/out/Debug/obj/gen/include -I../deps/v8/include -I../deps/icu-small/source/i18n -I../deps/icu-small/source/common -I../deps/openssl/openssl/include -I../deps/zlib -I../deps/http_parser -I../deps/cares/include -I../deps/uv/include  -O0 -gdwarf-2 -mmacosx-version-min=10.7 -arch x86_64 -Wall -Wendif-labels -W -Wno-unused-parameter -std=gnu++0x -stdlib=libc++ -fno-rtti -fno-exceptions -fno-threadsafe-statics -fno-strict-aliasing -MMD -MF /Users/jasongin/src/napi/napi-8/out/Debug/.deps//Users/jasongin/src/napi/napi-8/out/Debug/obj.target/node/src/node_api.o.d.raw   -c -o /Users/jasongin/src/napi/napi-8/out/Debug/obj.target/node/src/node_api.o ../src/node_api.cc

And that makes sense, because I don't see it defined anywhere in `node.gyp` or any `.gypi` that it includes.

My theory is this is related to how the linker resolves multiple definitions of inline functions,  which are not actually inlined in a debug build. One definition that _includes V8 checks_ comes from the V8 static lib, while a definition that _does not include checks_ comes from the compiled node sources. So the linker has to pick one when linking the executable. On Windows, the MSVC linker happens to pick the one that doesn't include the checks. On other platforms, we were just getting lucky (?) that the linker picked the V8 inline function definitions with the checks included.

addaleax replied:
@jasongin Is this enough to “fix” that for you on Windows?

```diff
diff --git a/common.gypi b/common.gypi
index 147cc70fa5d5..36f0dc760a8f 100644
--- a/common.gypi
+++ b/common.gypi
@@ -76,7 +76,7 @@
         'variables': {
           'v8_enable_handle_zapping': 1,
         },
-        'defines': [ 'DEBUG', '_DEBUG' ],
+        'defines': [ 'DEBUG', '_DEBUG', 'V8_ENABLE_CHECKS' ],
         'cflags': [ '-g', '-O0' ],
         'conditions': [
           ['target_arch=="x64"', {
```

jasongin replied:
Yes, I made exactly that change already locally and confirmed that enabled the V8 checks on Windows. I don't know if that's the right fix to submit to node though.

jasongin replied:
Fixed the functions that do number casting so they don't crash with `V8_ENABLE_CHECKS` in https://github.com/nodejs/node/pull/11975/commits/f21ab80825e62b302da65abbea49cff834650668

jasongin commented on src/node_api_internal.h:
```diff
@@ -0,0 +1,16 @@
+  v8::Isolate* V8IsolateFromJsEnv(napi_env e);
+
+  napi_value JsValueFromV8LocalValue(v8::Local<v8::Value> local);
+  v8::Local<v8::Value> V8LocalValueFromJsValue(napi_value v);
```
At one point these were also used in `node_api_async.cc` also, but it looks like they are not anymore. I'll see whether we can get rid of some or all of the `_internal.h` headers.

jasongin replied:
Resolved: There was no good reason for the `_internal.h` headers anymore, so I removed them in https://github.com/nodejs/node/pull/11975/commits/f21ab80825e62b302da65abbea49cff834650668

jasongin commented on doc/api/cli.md:
```diff
@@ -144,6 +144,13 @@ added: v6.0.0
 
+### `--napi-modules`
+<!-- YAML
+added: v8.0.0
```
Resolved by https://github.com/nodejs/node/pull/11975/commits/f21ab80825e62b302da65abbea49cff834650668

jasongin commented:
@addaleax I pushed another commit that addresses all your feedback from today. Thanks again!

richardlau commented on doc/node.1:
```diff
@@ -120,6 +120,10 @@ Throw errors for deprecations.
 
 .TP
+.BR \-\-napi\-modules
+Load N-API modules (experimental, opt-in by adding this flag).
```
This should match the description in `doc/api/cli.md` for consistency.

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

addaleax commented on src/node_api.h:
```diff
@@ -0,0 +1,477 @@
+#ifdef __cplusplus
+#define EXTERN_C_START extern "C" {
+#define EXTERN_C_END }
+#else /* ndef __cplusplus */
```
… `ndef`?

jasongin replied:
Removed these comments in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

addaleax commented on src/node_api.h:
```diff
@@ -0,0 +1,477 @@
+                                                    size_t* result);
+
+// Methods to coerce values
+// These APIs may execute user script
```
typo: `user scripts`?

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

addaleax commented on src/node_api_async.cc:
```diff
@@ -0,0 +1,69 @@
+    if (worker->work != NULL) {
+      delete reinterpret_cast<uv_work_t*>(worker->work);
+    }
+    delete worker;
```
is this mixing `malloc()` with `delete`? :/

mhdawson replied:
If we used malloc to create I think we should use free.

addaleax commented on src/node_api_async.cc:
```diff
@@ -0,0 +1,69 @@
+  uv_work_t* req = reinterpret_cast<uv_work_t*>(malloc(sizeof(uv_work_t)));
+  req->data = worker;
+  worker->work = req;
+  return reinterpret_cast<napi_work>(worker);
```
Should this clear the `execute/complete/destroy` fields by setting them to `nullptr`?

addaleax commented on src/node_api_async.h:
```diff
@@ -0,0 +1,20 @@
+NAPI_EXTERN void napi_async_set_data(napi_work w, void* data);
+NAPI_EXTERN void napi_async_set_execute(napi_work w, void (*execute)(void*));
+NAPI_EXTERN void napi_async_set_complete(napi_work w, void (*complete)(void*));
+NAPI_EXTERN void napi_async_set_destroy(napi_work w, void (*destroy)(void*));
```
Afaict none of these 3 are optional… why not pass them as parameters to `napi_create_async_work` and scrap these then?

addaleax replied:
This probably isn’t helpful feedback but I am not sure I’d use this API as a native module author … the libuv API for this is easier to use, accessible from C, and is not going away anytime soon

mhdawson replied:
The reason that the async APIs were included was to help on the abi stability front.  It seems to be a common use case for modules to need these async APIs. Nan provides an AsyncWorker class with similar functionality as does the C++ wrapper that will go along with the N-API, and we needed them for a number of the modules we ported. Without these APIs, if libuv changes any existing modules would break and need to be modified and recompiled.  I think that would be a surprise for those modules using the C++ wrapper as all other APIs used by the wrapper would be protected through the N-API.

addaleax commented on src/node_api_async.cc:
```diff
@@ -0,0 +1,69 @@
+
+void napi_async_queue_worker(napi_work w) {
+  napi_work_impl* worker = reinterpret_cast<napi_work_impl*>(w);
+  uv_queue_work(uv_default_loop(),
```
I guess using `uv_default_loop()` doesn’t play well with embedders using multiple Environments either?

addaleax commented on src/node_api_types.h:
```diff
@@ -0,0 +1,94 @@
+  void* data;
+} napi_property_descriptor;
+
+#define DEFAULT_ATTR 0, 0, 0, napi_default, 0
```
The macro name should start with `NAPI_`

jasongin replied:
Hmm, this isn't used anywhere, not even in tests. The use is also not very clear. I think it was left in as an oversight. I'm going to delete it.

jasongin replied:
Removed the macro in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

addaleax commented on test/addons-napi/2_function_arguments/binding.c:
```diff
@@ -0,0 +1,58 @@
+}
+
+#define DECLARE_NAPI_METHOD(name, func)                          \
+  { name, func, 0, 0, 0, napi_default, 0 }
```
Maybe this should go into the header? :smile:

jasongin replied:
The expectation is that most code will use the C++ wrapper classes that provide more convenient ways of doing this, so this sort of helper macros might not be widely needed. I'm not sure though; we can always add them later if it makes sense.

addaleax commented on test/addons-napi/3_callbacks/binding.c:
```diff
@@ -0,0 +1,51 @@
+#include <node_api.h>
+
+void RunCallback(napi_env env, const napi_callback_info info) {
```
There are 3 places where `const napi_callback_info` is used… why the `const`?

jasongin replied:
I don't know. I agree it serves no purpose, so I'll remove it.

(As mentioned at the top, this PR was a collaborative effort, so I didn't personally write or review every line.)

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

addaleax commented on test/addons-napi/3_callbacks/test.js:
```diff
@@ -0,0 +1,27 @@
+  assert.strictEqual(msg, 'hello world');
+});
+
+const global = function() { return this; }.apply();
```
Huh? This is strict mode, if you want the global object, just leave `global` the way it is.

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

addaleax commented on test/addons-napi/3_callbacks/test.js:
```diff
@@ -0,0 +1,27 @@
+    assert.strictEqual(this,
+                       (desiredRecv === undefined ||
+                        desiredRecv === null) ?
+                       global : desiredRecv);
```
These tests do fail, but the exception is swallowed? That’s not good.

jasongin replied:
It turns out this test was broken in multiple ways, so that it didn't test what it claimed to. I'm fixing it up. In this case, there was never any exception because the native callback encountered and ignored a `napi_invalid_arg`.

Normally, any exception thrown by JS code invoked by a native callback should get automatically propagated back up the JS stack, as long as it is not explicitly cleared by the native code.

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

I added consistent error checking for this test project using a `NAPI_CALL` macro. We have plans to move that macro into a shared header so it can be used by all the `addons-napi` tests, but that isn't done yet.

addaleax commented on test/addons-napi/4_object_factory/test.js:
```diff
@@ -0,0 +1,8 @@
+
+const obj1 = addon('hello');
+const obj2 = addon('world');
+assert.strictEqual(obj1.msg + ' ' + obj2.msg, 'hello world');
```
We like template literals for these cases ;)

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

addaleax commented on test/addons-napi/6_object_wrap/myobject.cc:
```diff
@@ -0,0 +1,199 @@
+    obj->env_ = env;
+    status = napi_wrap(env,
+                       jsthis,
+                       reinterpret_cast<void*>(obj),
```
You shouldn’t need a `reinterpret_cast` here, and iirc not even any cast at all?

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

Also in a couple other places that were doing `reinterpret_cast<void*>`

addaleax commented on test/addons-napi/7_factory_wrap/myobject.cc:
```diff
@@ -0,0 +1,107 @@
+MyObject::~MyObject() { napi_delete_reference(env_, wrapper_); }
+
+void MyObject::Destructor(void* nativeObject, void* /*finalize_hint*/) {
+  reinterpret_cast<MyObject*>(nativeObject)->~MyObject();
```
This seems sketchy … why isn’t this a `delete`? Where is the memory freed that’s being allocated by the `new MyObject()` call?

(Also, re: all the `reinterpret_cast`ing… we do try to avoid that if a `static_cast` does the job just fine)

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

addaleax commented on test/addons-napi/test_array/test.js:
```diff
@@ -0,0 +1,36 @@
+  test_array.Test(array, -2);
+} catch (err) {
+  assert.strictEqual(err.message, 'Invalid index. Expects a positive integer.');
+}
```
We have `assert.throws` method for exactly these cases :)

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

addaleax commented on test/addons-napi/test_buffer/test_buffer.c:
```diff
@@ -0,0 +1,156 @@
+
+static void noopDeleter(void *data, void* finalize_hint) {
+  (void)finalize_hint;
+  deleterCallCount++;
```
This could still at least check that `data == theText`, right?

jasongin replied:
Added an assert in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

This required adding a `napi_env` parameter to the `napi_finalize` callback, which after discussion we agreed was a good thing generally.

addaleax commented on test/addons-napi/test_buffer/test_buffer.c:
```diff
@@ -0,0 +1,156 @@
+
+void newBuffer(napi_env env, napi_callback_info info) {
+  napi_value theBuffer;
+  char *theCopy;
```
`char*`

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

addaleax commented on test/addons-napi/test_buffer/test_buffer.c:
```diff
@@ -0,0 +1,156 @@
+            napi_create_buffer(
+                env,
+                sizeof(theText),
+                (void **)(&theCopy),
```
We also try to avoid C-style casts, btw – if a proper cast pushes the line over 80 characters I personally would be okay, but this really doesn’t look like it does

jasongin replied:
Check the file extension: this is plain-C code.

Some of the tests are intentionally written in C, not C++, to validate that the ABI-stable API does not require C++.

addaleax replied:
Heh, that’s my bad then. Sorry!

addaleax commented on test/addons-napi/test_conversions/test_conversions.c:
```diff
@@ -0,0 +1,236 @@
+  status = napi_set_return_value(env, info, output);
+  if (status != napi_ok) goto Exit;
+
+Exit:
```
I don’t think we have uppercase labels elsewhere

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

addaleax commented:
Seems like Github is a bit overwhelmed with diffs of this size (I am too), let me know if you need help positioning my comments

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+  // via napi_define_class() can be (un)wrapped.
+  RETURN_STATUS_IF_FALSE(obj->InternalFieldCount() > 0, napi_invalid_arg);
+
+  obj->SetAlignedPointerInInternalField(0, native_object);
```
Also, should there be code here to handle the case that `native_object` is not an aligned pointer, like a `RETURN_STATUS_IF_FALSE()`? I would be okay with using an `External` to store the object, too.

jasongin replied:
Other places in the code use an `External` to store pointers in internal fields, this should too.

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/9c0b1517b894d0a655377d23f9e4b9943d93aa38

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+      : _this(this_arg), _args_length(args_length), _data(data) {}
+
+  virtual napi_value Holder() = 0;
+  virtual bool IsConstructCall() = 0;
```
Hmm, I'm not sure about `NewTarget()`, at least for now. I opened https://github.com/nodejs/abi-stable-node/issues/200 to track it.

I just noticed `Holder()` should actually be removed here; we had a `napi_get_cb_holder()` API but it was dropped for being too V8-specific also. (Though we may add it back if it is found to be necessary.)

jasongin replied:
Removed `Holder()` in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878.

We'll consider exposing the new target later, separate from this PR.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+  void* _data;
+};
+
+template <typename Info, int InternalFieldIndex>
```
A `k` prefix would imply to me that it is a constant. Since this is a template parameter, I think the TitleCase with no `k` prefix is appropriate. Unless you are aware of some other precedent in the node codebase?

addaleax replied:
@jasongin I know there’s precedent because I receveid the same suggestion in https://github.com/nodejs/node/pull/6357#r60899766 :smile:

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+    cbdata->SetInternalField(
+        v8impl::kGetterIndex,
+        v8::External::New(isolate, reinterpret_cast<void*>(getter)));
+  }
```
I'm pretty sure it is valid to define a setter without a getter. Highly unusual, but valid.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+  v8::Local<v8::FunctionTemplate> tpl = v8::FunctionTemplate::New(
+      isolate, v8impl::FunctionCallbackWrapper::Invoke, cbdata);
+
+  return_value = scope.Escape(tpl->GetFunction());
```
Do you know if there is a way to get warnings or something for any use of APIs that are tagged with `V8_DEPRECATE_SOON`?

addaleax replied:
@jasongin What I’ve done is to just cd into `out`, take the compile command that ran with a normal `make` and add `-DV8_IMMINENT_DEPRECATION_WARNINGS=1` manually… that worked well enough for me

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

I also built with the `V8_IMMINENT_DEPRECATION_WARNINGS` flag to ensure there weren't any other V8 calls that needed similar updates.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+
+      tpl->PrototypeTemplate()->SetAccessor(
+        property_name,
+        v8impl::GetterCallbackWrapper::Invoke,
```
I think it's valid for the getter to be null if the setter is not null.

Yes, I'll switch to the `Maybe` overload of `SetAccessor()`.

addaleax replied:
@jasongin I didn’t check that, but… you’re always passing `v8impl::GetterCallbackWrapper::Invoke` here, so no matter whether `p->getter` was set or not V8 is going to call `Invoke()` as the getter, which will try to call out to the callback (which will crash because it’s missing).

So maybe this should be `p->getter ? v8impl::GetterCallbackWrapper::Invoke : nullptr`, too?

jasongin replied:
Actually there is not any overload of `ObjectTemplate::SetAccessor()` that returns a `Maybe`.

addaleax replied:
@jasongin Oops, mixed this up with `Object::SetAccessor`. Sorry!

jasongin replied:
Added checks for null `p->getter` in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_get_propertynames(napi_env env,
```
No, I agree it should have a `_`. (I think I know why it was originally written this way, but that's not relevant anymore.)

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_get_boolean(napi_env env, bool value, napi_value* result) {
```
The thinking was that the `undefined`, `null`, `true`, and `false` values are conceptually singleton values: you don't ever actually create an instance of one of them you just get a handle. I guess whether that's actually true or not depends on the VM implementation, since they are not cached at the N-API layer. Personally I'd be fine with either name, and I don't think anyone feels strongly about it.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+
+  for (size_t i = 0; i < argc; i++) {
+    args[i] = v8impl::V8LocalValueFromJsValue(argv[i]);
+  }
```
Yes, good point.

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+  if (try_catch.HasCaught()) {
+    return napi_set_last_error(napi_pending_exception);
+  } else {
+    CHECK_MAYBE_EMPTY(maybe, napi_generic_failure);
```
I'm not sure, but I guess if the arguments to `Call()` were invalid in some way so that the function could not even be called? It seems safest to check it anyway.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+
+// Attempts to get a referenced value. If the reference is weak, the value might
+// no longer be available, in that case the call is still successful but the
+// result is NULL
```
I think there should probably be separate APIs for `DataView`. I opened https://github.com/nodejs/abi-stable-node/issues/202 to track that.

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin replied:
I opened https://github.com/nodejs/abi-stable-node/issues/209 to track this. Since it is purely an optimization that does not affect the public API surface, I think we will wait until later to implement this.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+    }
+
+    if (delete_self) {
+      delete reference;
```
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+    napi_status status =
+        napi_define_properties(env,
+                               *result,
+                               static_cast<int>(static_descriptors.size()),
```
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+      // should we return a different error code if the define failed?
+      if (define_maybe.IsNothing() || !define_maybe.FromMaybe(false)) {
+        return napi_set_last_error(napi_generic_failure);
+      }
```
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878 in both places.

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+      auto set_maybe = obj->SetAccessor(
+          context,
+          name,
+          v8impl::GetterCallbackWrapper::Invoke,
```
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

jasongin commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+  CHECK_ARG(result);
+
+  v8::Local<v8::Object> obj =
+      v8impl::V8LocalValueFromJsValue(js_object).As<v8::Object>();
```
Fixed in https://github.com/nodejs/node/pull/11975/commits/e1ca374cb5be63f6f5167bd4e6b40a01f9ecb878

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2515 @@
+  // via napi_define_class() can be (un)wrapped.
+  RETURN_STATUS_IF_FALSE(obj->InternalFieldCount() > 0, napi_invalid_arg);
+
+  *result = v8::Local<v8::External>::Cast(obj->GetInternalField(0))->Value();
```
Can you `RETURN_STATUS_IF_FALSE` if the internal field is not an `v8::External`?

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/20b6cf24dda5f7c3be607b08021dc8a05a36c429

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2478 @@
+                                  napi_ref* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+  RETURN_STATUS_IF_FALSE(initial_refcount >= 0, napi_invalid_arg);
```
@jasongin Any thoughts? :)

jasongin replied:
Fixed in https://github.com/nodejs/node/pull/11975/commits/375af79b23ed02c3a46a3b28f12a4d6eb3721990

jasongin commented:
@addaleax There are a few issues we're still working on:
 - Changing the signed refcount to an unsigned int
 - Caching `Symbol.hasInstance`
 - Async work APIs: after further review we found these are in need of significant redesign, to incorporate feedback and add error handling consistent with the other APIs. We're still discussing how we'll approach that. See https://github.com/nodejs/abi-stable-node/pull/204#issuecomment-289888995

addaleax commented:
@jasongin Yeah, I just meant to say that pending the currently unresolved issues I should be able to approve this :) The first two items shouldn’t be blockers for the PR anyway.

mhdawson commented:
@addaleax thanks for your quick and through reviews.

jasongin commented:
I pushed 4 more commits today:

 1. Switch to unsigned int refcount: https://github.com/nodejs/node/pull/11975/commits/375af79b23ed02c3a46a3b28f12a4d6eb3721990
 2. Add a check for external value type in `napi_unwrap`: https://github.com/nodejs/node/pull/11975/commits/20b6cf24dda5f7c3be607b08021dc8a05a36c429
3. Add `napi_env` parameter to `napi_get_last_error_info()`: https://github.com/nodejs/node/pull/11975/commits/232f00454d83c20e3015d919aeb24af5696f82d0 - Not based on any feedback above, but while reviewing we found this was an oversight, and thought it better to make this breaking change sooner than later. This is a prerequisite for the work at https://github.com/nodejs/node/pull/11975/commits/232f00454d83c20e3015d919aeb24af5696f82d0, but we decided to wait on that internal refactoring until after this PR lands, because it isn't critical right now.
4. _Temporarily_ remove the async work APIs, because they are not ready yet. They still need more development, and testing, and we didn't want that to be rushed or to block this PR. For further explanation refer to https://github.com/nodejs/abi-stable-node/pull/204#issuecomment-290203050

@addaleax and others reviewers: With these updates I think we have addressed all the feedback on this PR so far.

jasongin commented on test/addons-napi/3_callbacks/binding.c:
```diff
@@ -0,0 +1,49 @@
+#include <node_api.h>
+
+#define NAPI_CALL(theCall)                                                \
+  if (theCall != napi_ok) {                                               \
```
Fixed in https://github.com/nodejs/node/pull/11975/commits/20b6cf24dda5f7c3be607b08021dc8a05a36c429

addaleax commented on src/node_api.cc:
```diff
@@ -0,0 +1,2530 @@
+// After this call the reference will be a strong reference because its
+// refcount is >0, and the referenced object is effectively "pinned".
+// Calling this when the refcount is 0 and the object is unavailable
+// results in an er
```
@boingoing Not sure about this comment – there are no calls into the VM here, so this shouldn’t be a problem?

addaleax replied:
Eep, Github messing with the diff again. This should refer to the `TODO(boingoing)` in `napi_get_and_clear_last_exception`.

boingoing replied:
At first I was confused about the line. :)

This code comment is pretty old at this point, I think the answer is no and we should remove it.

jasongin replied:
Removed in https://github.com/nodejs/node/pull/11975/commits/aa22a2ab1c7fa006b35eb155174334e6958b7922

addaleax commented on src/node_api_types.h:
```diff
@@ -0,0 +1,95 @@
+  napi_int32,
+  napi_uint32,
+  napi_float32,
+  napi_float64,
```
I’d suggest adding `_array` suffixes… these end up in the global namespace, and without context `napi_uint32` sounds more like a datatype then an enum value

jasongin replied:
Yes, sounds like a good idea.

jasongin replied:
Added in https://github.com/nodejs/node/pull/11975/commits/aa22a2ab1c7fa006b35eb155174334e6958b7922

jasongin commented:
> I’d prefer for including the changes that are more or less sure to happen and touch the public headers here (nodejs/abi-stable-node#208 in particular) before landing, though.

The public header part of that is included as https://github.com/nodejs/node/pull/11975/commits/232f00454d83c20e3015d919aeb24af5696f82d0. The part left for later is only internal refactoring (to support multiple isolates/environments). The PR at https://github.com/nodejs/abi-stable-node/pull/208 hasn't yet been updated to reflect that separation.

addaleax commented:
New CI: https://ci.nodejs.org/job/node-test-commit/8760/

Looks like you need to take care of a linter issue for a recently added rule:

```
/home/sqrt/src/node/test/addons-napi/test_array/test.js
  29:16  error  Unescaped dot character in regular expression  no-unescaped-regexp-dot
  29:44  error  Unescaped dot character in regular expression  no-unescaped-regexp-dot
```

That’s so minor I’d say it could happen while landing the PR if necessary.

jasnell commented:
I am *temporarily* marking this as don't land in v7, v6 or v4. There is every intent to get this backported but this specific PR should not be landed in those at this time, at least not until we've worked through all the issues.

addaleax commented:
How about we say I land it on Monday if nobody else here shows that they have the intention of giving this a full review (and the time do so in the foreseeable future) by then? I agree, getting more reviews wouldn’t hurt, but so far it looks a little dim. :smile:

@nodejs/abi-stable-node does that work for you?

gabrielschulhof commented:
@addaleax I have now rebased https://github.com/nodejs/abi-stable-node/pull/208 onto the latest in the api-prototype-8.x branch, so now it's purely an internal refactor, touching only src/node_api.cc.

aruneshchandra commented:
@addaleax that sounds like a good idea and should totally work for us. People who would like to review this and need more time should let this thread know about their intentions before Monday, so that we can hold it longer for them.

kunalspathak commented:
Started new CI - https://ci.nodejs.org/job/node-test-pull-request/7108/

mhdawson commented:
I'd sign off as another CTC member, but as one of the contributors to the PR I can't do that.  But just saying..

addaleax commented on src/node_api_types.h:
```diff
@@ -0,0 +1,95 @@
+                              napi_callback_info info);
+typedef void (*napi_finalize)(napi_env env,
+                              void* finalize_data,
+                              void* finalize_hint);
```
These should be `extern "C"`, right? Maybe they should also get `NAPI_EXTERN`, and `NAPI_EXTERN` could get `extern "C"` prepended… or maybe even `extern "C" __cdecl` to be even more precise?

jasongin replied:
`NAPI_EXTERN` is not appropriate here because typedefs are not symbols that get exported or imported.

My understanding is that `extern "C"` language linkage controls two things: name mangling and calling convention. Name mangling is not an issue because there are no exported symbols for the typedefs. So then the only issue is calling convention. I thought `__cdecl` would be the default for both C and C++, but maybe it is different for some platforms/compilers that node supports?

addaleax replied:
> `NAPI_EXTERN` is not appropriate here because typedefs are not symbols that get exported or imported.

That’ true, but you also have that in `node_api.h`. I’d remove it in both places then.

> I thought `__cdecl` would be the default for both C and C++, but maybe it is different for some platforms/compilers that node supports?

Mh, I think the x64 default calling conventions don’t match __cdecl. It’s been a while since I’ve had to deal with this stuff… leaving it this way should be fine.

jasongin replied:
> You also have that in node_api.h. I'd remove it in both places then.

I don't understand. NAPI_EXTERN is used on exported functions there, not typedefs.

And yeah, calling convention is different with x64, but then it shouldn't be affected by extern "C" linkage as far as I know.

addaleax replied:
@jasongin What about `napi_addon_register_func`?

jasongin replied:
Oh, good point! That one does look like a mistake. I don't see any reason why `napi_addon_register_func` should have `NODE_EXTERN`. (It should be declared the same way as `addon_register_func` in `node.h`.)

jasongin replied:
Fixed as https://github.com/nodejs/node/pull/11975/commits/c87eade56391fe3db6bbc1c17d7f449f0599bb46

Fishrock123 commented:
Kinda wish it was a bit more like the [duktape api](http://duktape.org/api.html)* which is quite similar in API shape generally, but returns things from functions you;d expect to return things, rather than the status. I do not know if that is realistic or not.

(* imo duktape has a very nice C api)

kkoopa commented:
> Kinda wish it was a bit more like the duktape api* which is quite similar in API shape generally, but returns things from functions you;d expect to return things, rather than the status.

From a glance, it appears that duktape is implemented as a stack machine, so it is _very_ different.

addaleax commented:
Landed in 56e881d0b0b2997b518753cb627eb3b50eeb6f62 :tada:

@nodejs/abi-stable-node I think that also means that from now on you can open PRs against this repository instead of nodejs/abi-stable-node. :)
